src/runlib.rs

```diff
@@ -1 +1,196 @@
+                let path = entry_path.to_str().unwrap().to_string().replace("./", "");
+                artifacts.insert(VirtualTargetPath::new(path)?, hashes);
+            }
+        }
+    }
+    Ok(artifacts)
+}
+
+/// run_command is a function that, given command arguments, executes commands on a software supply chain step
+/// and returns the stdout and stderr as byproducts.
+/// The first element of cmd_args is used as executable and the rest as command arguments.
+/// If a commands in run_command fails to execute, the error is returned.
+pub fn run_command(cmd_args: &[&str], run_dir: Option<&str>) -> Result<BTreeMap<String, String>> {
+    let executable = cmd_args[0];
+    let args = (&cmd_args[1..])
+        .iter()
+        .map(|arg| {
+            if VirtualTargetPath::new((*arg).into()).is_ok() {
+                let absolute_path = canonicalize_path(*arg);
+                match absolute_path {
+                    Ok(path_buf) => match path_buf.to_str() {
+                        Some(p) => p,
+                        None => *arg,
+                    },
+                    Err(_) => *arg,
+                };
+            }
+            *arg
+        })
+        .collect::<Vec<&str>>();
+
+    let mut cmd = Command::new(executable);
+    let mut cmd = cmd.args(args);
+
+    if let Some(dir) = run_dir { cmd = cmd.current_dir(dir) }
+
+    let output = cmd.output()?;
+
+    // Emit stdout, stderror
+    io::stdout().write_all(&output.stdout)?;
+    io::stderr().write_all(&output.stderr)?;
+
+    // Format output into Byproduct
+    let mut byproducts: BTreeMap<String, String> = BTreeMap::new();
+    // Write to byproducts
+    let stdout = match String::from_utf8(output.stdout) {
+        Ok(output) => output,
+        Err(error) => {
+            return Err(Error::from(io::Error::new(
+                std::io::ErrorKind::Other,
+                format!("Utf8Error: {}", error),
+            )))
+        }
+    };
+    let stderr = match String::from_utf8(output.stderr) {
+        Ok(output) => output,
+        Err(error) => {
+            return Err(Error::from(io::Error::new(
+                std::io::ErrorKind::Other,
+                format!("Utf8Error: {}", error),
+            )))
+        }
+    };
+    let status = match output.status.code() {
+        Some(code) => code.to_string(),
+        None => "Process terminated by signal".to_string(),
+    };
+
+    byproducts.insert("stdout".to_string(), stdout);
+    byproducts.insert("stderr".to_string(), stderr);
+    byproducts.insert("return-value".to_string(), status);
+
+    Ok(byproducts)
+}
+
+// TODO: implement default trait for in_toto_run's parameters
+
+/// in_toto_run is a function that executes commands on a software supply chain step
+/// (layout inspection coming soon), then generates and returns its corresponding Link metadata.
+pub fn in_toto_run(
+    name: &str,
+    // run_dir: Option<&str>,
+    material_paths: &[&str],
```

Comment:
Should we make the path arguments `&[&str]` or `Vec<String>` (or maybe `Vec<VirtualTargetPath>`?)

Reply:
Same with `name` and other params (&str vs String)

Reply:
After a discussion with @adityasaky, we decided to stick with more accessible types like string slices as opposed to custom types such as `VirtualTargetPath` for function parameters (similar to other in-toto implementations) and handle those conversions internally.